### PR TITLE
Implement stochastic gradient descent with warm restarts

### PR DESCRIPTION
**Pull request by jveitchmichaelis**
I've drafted an implementation of (cosine) stochastic gradient descent with warm restarts: https://arxiv.org/abs/1608.03983

* Adds a user-defined minimum learning rate
* Adds a user-defined cycle length (in batches)
* SGDR is added as a learning rate policy choice:

```c
case SGDR:
        rate = net.learning_rate_min + 
                    0.5*(net.learning_rate-net.learning_rate_min)
                    * (1. + cos( (float) (batch_num % net.batches_per_cycle)*3.14159265 / net.batches_per_cycle));
      
        return rate;
```

Some thoughts:

* Might be useful to build in learning rate decay - e.g. reduce net.learning_rate_max as with the step policy.
* Could also implement a multiplication factor to increase cycle length over time.
* Would it make more sense to use a cycle length based on epoch rather than batches? 
* Would be interesting to do some testing to see if and how much it improves training object detectors, as it's an increasingly popular approach for classification networks (I believe fast.ai use this as their default scheduler now).

### CODE REVIEW

**Comment by AlexeyAB** [comment]
@jveitchmichaelis Hi, Thanks for PR

Did you try to use it for training and did you get some acceleration of training or increased accuracy?
Why you didn't use `learning_rate_max` ?
Are these default values: lr_min=.00001 and sgdr_cycle=500 are optimal?

**Comment by jveitchmichaelis** [comment]
I've successfully trained a couple of networks using this. I found it worked fairly well (I got convergence at least), though I think combining with the STEPS variable to decay the learning rate after 8k/10k iterations is helpful. In other implementations typically the minimum learning rate you decay to is a couple of orders of magnitude lower than the starting point, so it seemed like a good place to start.

I wasn't sure about a separate `learning_rate_max` variable, since there is already a `learning_rate` option which is effectively the same thing.

**Comment by AlexeyAB** [comment]
Thanks for testing. Did you get convergence faster than by using common way, or did you final loss lower than by using common way?

**Comment by LukeAI** [comment]
Looks like a great addition! Does it improve loss and/or convergence? I'm happy to test and report back - how do I enable it?

**Comment by jveitchmichaelis** [comment]
You need to set the policy in your config file to SGDR and also add a `learning_rate_min` and a `sgdr_cycle` which the rate will oscillillate between. There are defaults for these two, so minimally just setting the policy will do.

I found it worked reasonably well on my dataset, which was jumping around a lot with Steps. It would be good to check what cycle length is good as a default, I'm not sure what the theory is there.

**Comment by jveitchmichaelis** [comment]
@LukeAI you need to change the policy in your config file to `SGDR`. You can also add a cycle length and `learning_rate_min` (these have defaults though so not necessary). You might also want to play with the burn in time , I set mine to 500.
